ext/integrations/exec_integration.c

```diff
@@ -0,0 +1,365 @@
+    FREE_HASHTABLE(tracked_streams);
+    tracked_streams = NULL;
+}
+
+static PHP_FUNCTION(DDTrace_integrations_exec_register_stream) {
+    php_stream *stream;
+    zval *zstream;
+    zend_object *span;
+
+    ZEND_PARSE_PARAMETERS_START(2, 2)
+    Z_PARAM_RESOURCE(zstream)
+    Z_PARAM_OBJ(span)
+    ZEND_PARSE_PARAMETERS_END();
+
+    php_stream_from_res(stream, Z_RES_P(zstream));
+    if (!stream) {
+        RETURN_FALSE;
+    }
+
+    dd_exec_lazy_init_track_streams();
+
+    {
+        zval zspan;
+        ZVAL_OBJ(&zspan, span);
+        zend_hash_str_add(tracked_streams, (const char *)&stream, sizeof stream, &zspan);
+        GC_ADDREF(span);
+    }
+
+    RETURN_TRUE;
+}
+
+static int dd_php_stdiop_close_wrapper(php_stream *stream, int close_handle) {
+    int ret = orig_php_stream_stdio_ops_close(stream, close_handle);
+
+    if (!in_request) {
+        return ret;
+    }
+
+    dd_exec_lazy_init_track_streams();
+    zval *span_data_zv = zend_hash_str_find(tracked_streams, (const char *)&stream, sizeof stream);
+    if (!span_data_zv) {
+        return ret;
+    }
+
+    ddtrace_span_data *span_data = OBJ_SPANDATA(Z_OBJ_P(span_data_zv));
+
+    zend_array *meta = ddtrace_property_array(&span_data->property_meta);
+    if (ret == -1) {
+        zval zv;
+        ZVAL_INTERNED_STR(&zv, pclose_minus_one_zstr);
+        zend_hash_update(meta, error_message_zstr, &zv);
+    } else {
+        zval zexit;
+        ZVAL_LONG(&zexit, ret);
+        zend_hash_update(meta, cmd_exit_code_zstr, &zexit);
+    }
+
+    dd_trace_stop_span_time(span_data);
+    ddtrace_close_span_restore_stack(span_data);
+
+    zend_hash_str_del(tracked_streams, (const char *)&stream, sizeof stream);
+
+    return ret;
+}
+
+/* proc_open / proc_close handling */
+typedef struct _dd_proc_span {
+    zend_object *span;
+    php_process_id_t child;
+} dd_proc_span;
+static int le_proc;
+static int le_proc_span;
+
+static void dd_proc_wrapper_rsrc_dtor(zend_resource *rsrc) {
+    // this is called from the begginning of proc_open_rsrc_dtor,
+    // before the process is possibly reaped
+
+    dd_proc_span *proc_span = (dd_proc_span *)rsrc->ptr;
+
+    ddtrace_span_data *span_data = OBJ_SPANDATA(proc_span->span);
+
+    if (span_data->duration == 0) {
+        // if we reaped, the proc handle destructor will set
+        // FG(pclose_ret) to -1, causing proc_close to return -1, so we
+        // must signal that we have to override the return value in the post
+        // hook of proc_close
+        bool did_reap = dd_waitpid(span_data, proc_span->child);
+        if (did_reap) {
+            // set a dynamic property on the span
+#if PHP_VERSION_ID <= 80000
+            zval zobj;
+            ZVAL_OBJ(&zobj, proc_span->span);
+            zend_update_property_bool(ddtrace_ce_span_data, &zobj, ZEND_STRL("overrideRetval"), 1);
+#else
+            zend_update_property_bool(ddtrace_ce_span_data, proc_span->span, ZEND_STRL("overrideRetval"), 1);
```

Comment:
Newer PHP versions aren't going to be happy with dynamic properties...
But I think we actually don't _need_ to check for overrideRetval - just check in proc_close pre-hook whether exit status is already set, and if it is, do nothing, otherwise assign it to the new value.

Reply:
Updated accordingly